### PR TITLE
fix: prevent concurrent registration of client-side routes

### LINKED ISSUE #2329: Duplicate key detected in ClientRouteRegistry
### Describe the bug

I randomly get the runtime exception about duplicate key when Hilla tries to get all routes from `ClientRouteRegistry`:

```
java.lang.IllegalArgumentException: duplicate key: /about
        at java.base/java.util.ImmutableCollections$MapN.<init>(ImmutableCollections.java:1196) ~[na:na]
        at java.base/java.util.Map.ofEntries(Map.java:1680) ~[na:na]
        at java.base/java.util.Map.copyOf(Map.java:1748) ~[na:na]
        at com.vaadin.hilla.route.ClientRouteRegistry.getAllRoutes(ClientRouteRegistry.java:67) ~[hilla-endpoint-24.4.0.alpha21.jar:na]
        at com.vaadin.hilla.route.RouteUnifyingIndexHtmlRequestListener.collectClientViews(RouteUnifyingIndexHtmlRequestListener.java:115) ~[hilla-endpoint-24.4.0.alpha21.jar:na]
        at com.vaadin.hilla.route.RouteUnifyingIndexHtmlRequestListener.modifyIndexHtmlResponse(RouteUnifyingIndexHtmlRequestListener.java:82) ~[hilla-endpoint-24.4.0.alpha21.jar:na]

```

This prevents application to start.
Happens in both dev and prod modes, but quite randomly.

### Expected-behavior

No duplicates in routes and no exception on startup.

### Reproduction

This can be reproduced with the branch https://github.com/vaadin/flow-hilla-hybrid-example/pull/29.

Download the starter, checkout the branch and start with `mvn` (or in production mode `mvn spring-boot:run -Pproduction`).

Reproduces randomly locally and in GHA. You need to try several times.

### System Info

Vaadin 24.4.0.alpha21, Hilla 24.4.0.alpha21

### PR DESCRIPTION
**Pull request by taefi**
The approach is to synchronize all the methods that can
manipulate the registeredRoute map. Also, Loading the
file-routes.json in production mode moved to the related
serviceInitListener which happens once at startup before
and requests could be served. As that's the only time
client-side routes can/should be registered, the risk of
concurrent registration would be zero. Then, in development
mode, the chances of getting concurrent requests passing
through the indexhtml listener is low by nature, however
due to the synchronization practices introduced in this
commit, there should not be a situation that allows any
concurrency in dev mode either.

Fixes #2329

### CODE REVIEW

**Comment by taefi** [comment]
> What problem does this solve? Is it intentional to no longer persevere ordering?

I don't believe the previously ordered map was serving any purpose. The `createMenuItems` utility is reading the values from `window.Vaadin.server.views` and sorts it based on its own logic.

There seems to be rare cases of race condition when reading the `file-routes.json` and adding entries with duplicate keys to this Map, and it is supposed to solved that.

**Comment by Legioth** [comment]
How can we be sure that we're just not shifting the problem to some other part of the stack if we don't first understand what causes the issue in the first place?

**Comment by Legioth** [comment]
In other words: if we assume that the underlying map is being modified concurrently, then a `ConcurrentHashMap` will help us make the `copyOf` capture a consistent snapshot but we cannot know whether that snapshot is always the right snapshot.

To really fix the issue, we need to understand why that happens so that we can identify the right moment to read a snapshot. One that moment is identified, we can determine whether a `ConcurrentHashMap` is the appropriate tool for achieving that.

**Comment by taefi** [comment]
> we can identify the right moment to read a snapshot.

TBH, I didn't understand the reasoning behind the above, in other words I want to ask why should we care about the "when"? To my understanding, even a simple Map should never have duplicate keys, so we need to prevent the concurrent modification of that map, and to do so, using the ConcurrentHashMap may or may not be sufficient. 

 Those failing tests that revealed the issue are running in production mode. Looking at the code, [this](https://github.com/vaadin/hilla/blob/main/packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java#L126) is the culprit, since the tests can run in parallel and this could be hit by multiple threads at a time.

**Comment by cromoteca** [comment]
I don't get which tests are run in parallel. It could be useful to amend the `addRoute` method to avoid adding a route twice. That should trigger an immediate error.

**Comment by mshabarov** [comment]
> I don't get which tests are run in parallel. It could be useful to amend the `addRoute` method to avoid adding a route twice. That should trigger an immediate error.

I've just reproduced the issue with `mvn verify -Pit,production`, which starts the app and opens several browsers, then I get as much exceptions as opened browsers.

**Review by cromoteca — DISMISSED** [review]
The problem of this PR is that it masks an issue rather than solving it.

Adding a check for existing routes in the `addRoute` method would disallow any code that tries to overwrite routes. Then, we'd need to understand if we need to fix the offending code or our API.

**Comment by taefi** [comment]
Maybe my mention of the following piece of code in an earlier comment wasn't visible enough: 

https://github.com/vaadin/hilla/blob/main/packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java#L126

Multiple browsers requesting the index.html at the same time (in production mode) would result in concurrent attempts for calling the `collectClientViews` method, and since the `lastUpdated` is not getting updated until the `ClientRouteRegistry::registerClientRoutes` finishes reading the file-routes.json and add entires to the `registeredRoutes` map, the race condition is happening there. Marking `lastUpdated` as `volatile`, in conjunction with the synchronizing the `collectClientViews` will prevent this issue from happening. 

However, since `ClientRouteRegistry::addRoute` is a public method, it could be called from other places, so the strategy of "what should be done when another code attempts to register duplicates" is another question to answer. We don't want duplicates for sure, deciding between `put` to overwrite, or `putIfAbsent` to prevent overwrites is up to us to decide. We can prevent the overwrites by throwing an exception to preventing the app from getting any further, just in case.

**Comment by cromoteca** [comment]
I'd throw an exception if a duplicate route is added.

Concerning `lastUpdated`, I'd rather write its value immediately, then perform the update, eventually clearing the value if the update goes wrong for any reason.

**Comment by Legioth** [comment]
> why should we care about the "when"?

If there's a duplicate, then it sounds like sometimes key X maps to value A and sometimes to value B. Should we use A or B?

**Comment by taefi** [comment]
> If there's a duplicate, then it sounds like sometimes key X maps to value A and sometimes to value B. Should we use A or B?

I believe two entries sharing the same key is the result of a bad config in the user application. Throwing an error with an informative message when the duplication attempt is made is good enough to inform the user about the existence of a bad config in their project.

**Comment by cromoteca** [comment]
As @taefi explained above, all the views are being inserted twice, with different outcome given that the collection wasn't thread-safe. So it's the same view getting added twice. This happens quite easily since there's a long time-frame between checking `lastUpdated` and setting it.
While discussing the issue, I proposed to synchronize on the registry class instead than on the collection, so that the whole update process cannot be done twice at the same time and the second thread has to wait for the result to come.

**Comment by mshabarov** [comment]
I did a bit of testing by adding an extra logging line into `registerAndRecurseChildren` method at the beginning, then run the IT tests.

Then I clearly saw that this logging message was recorded N x M times, where N is a number of routes in the app and M is a number of opened browser windows.

This isn't a matter of having different routes A and B for the same path key (at least in most of the cases), it's a matter of concurrent modification of registered routes collection.

Would it work if for the production mode Hilla calculates this collection in advance, even before bootstrap request from clients and for dev mode it can calculate on demand by client (to have live reload for routes changes), but in a synchronized way?

**Comment by taefi** [comment]
We can definitely trigger loading and initializing the `ClientRouteRegistry` for the production mode in the `RouteUnifyingServiceInitListener`, since the `file-routes.json` is already created during the build. 
Probably, we can also move the `lastUpdated` variable to the `ClientRouteRegistry` and make the operations related to loading/reloading the `file-routes.json` isolated and atomic in that class. Though, I need to verify that whether is it possible for a request to reach the IndexHtmlListener before the `RouteUnifyingServiceInitListener` finishes its job or not.

**Review by platosha — APPROVED** [review]

File: packages/java/endpoint/src/main/java/com/vaadin/hilla/route/ClientRouteRegistry.java [hunk 1]
```diff
@@ -58,42 +60,42 @@ public class ClientRouteRegistry implements ClientRoutesProvider {
...
-     * Removes a route from the registry.
+     * @throws IllegalStateException
+     *             if the route has already been registered
      *
-     * @param route
-     *            the route to remove
      */
-    public void removeRoute(String route) {
-        registeredRoutes.remove(route);
+    synchronized void addRoute(String route, ClientViewConfig clientView) {
```
[the remark this answers is not in the document]
Comment (taefi) [reply packages/java/endpoint/src/main/java/com/vaadin/hilla/route/ClientRouteRegistry.java:92]: Well, technically they aren't fully private, that's why I added the `synchronized` :)

**Review by mshabarov — APPROVED** [review]

File: packages/java/endpoint/src/main/java/com/vaadin/hilla/route/ClientRouteRegistry.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mshabarov) [reply packages/java/endpoint/src/main/java/com/vaadin/hilla/route/ClientRouteRegistry.java:92]: This shouldn't lead to any harm though, let's keep them.

**Pull request merged**